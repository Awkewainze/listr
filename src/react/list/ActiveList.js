import React from 'react';
import './ActiveList.scss';
import ListItem from './item/ListItem';
import TransitionGroup from 'react-transition-group/cjs/TransitionGroup';
import AppCSSTransition from '../common/AppCSSTransition';

const ActiveList = ({ items, onDelete, onToggle }) => (
  <TransitionGroup className="ActiveList">
    {items.map((item, i) =>
      <AppCSSTransition key={item.id} prefix="ListItem" timeout={200}>
        <ListItem
          key={item.id}
          item={item.text}
          checked={item.checked}
          onDelete={onDelete}
          index={i}
          itemId={item.id}
          onCheck={onToggle}
        />
      </AppCSSTransition>
    )}
  </TransitionGroup>
);

export default ActiveList;