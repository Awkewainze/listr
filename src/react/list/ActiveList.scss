@import '../../variables';
@import '../../mixins';

.ActiveList {
  .ListItem {
    transition: opacity 200ms ease;

    @include transitionState($exiting) {
      opacity: 0;
    }
  }
}