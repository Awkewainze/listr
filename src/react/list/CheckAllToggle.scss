@import "../../mixins";

.CheckAllToggle {
  @include gridLayout();
  grid-column: 1 / 5;
  align-items: center;
  margin: 12px 0;

  .Checkbox {
    grid-column: 2 / 3;
  }

  .CheckAllToggle__textWrap {
    grid-column: 3 / 4;

    .CheckAllToggle__textLabel {
      font-style: italic;
      font-size: 14px;
      cursor: pointer;
      user-select: none;
    }
  }
}