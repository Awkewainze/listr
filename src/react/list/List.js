import React from 'react';
import './List.scss';
import TransitionGroup from 'react-transition-group/cjs/TransitionGroup';

/*
 * This component will throw a warning during deletion of items.
 * See https://github.com/reactjs/react-transition-group/issues/429
 * and https://github.com/reactjs/react-transition-group/issues/606
 * We need to wait for react-transition-group to come to a solution to resolve this.
 */
const List = ({ children }) => (
  <TransitionGroup className="ActiveList">
    {children}
  </TransitionGroup>
);

export default List;