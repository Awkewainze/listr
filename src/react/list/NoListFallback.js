import React from 'react';

const NoListFallback = () => (
  <div className="NoListFallback">
    <h3>This is unexpected.</h3>
    <p>We're sorry, we're not sure how you got here! No list could be loaded.</p>
  </div>
);

export default NoListFallback;