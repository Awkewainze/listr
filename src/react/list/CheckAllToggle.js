import React from 'react';
import './CheckAllToggle.scss';
import Checkbox from '../common/Checkbox';
import { useSelector, useDispatch } from 'react-redux';
import selectActiveListAllChecked from '../../redux/selectors/list/selectActiveListAllChecked';
import selectActiveListId from '../../redux/selectors/list/selectActiveListId';
import { toggleAllChecked } from '../../redux/actions/list';

const CheckAllToggle = ({ id }) => {
  const activeList = useSelector(selectActiveListId);
  const allChecked = useSelector(selectActiveListAllChecked);
  const dispatch = useDispatch();

  const onChange = () => {
    dispatch(toggleAllChecked(activeList));
  };

  return (
    <div className="CheckAllToggle">
      <Checkbox
        checked={allChecked}
        id={id}
        onChange={onChange}
      />
    </div>
  );
};

export default CheckAllToggle;