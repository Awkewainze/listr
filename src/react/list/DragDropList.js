import React from 'react';
import { useDispatch } from 'react-redux';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import List from './List';
import AppCSSTransition from '../common/AppCSSTransition';
import ListItem from './item/ListItem';
import { reorderItem } from '../../redux/actions/list';

const DragDropList = ({
  items,
  listId,
  onDelete,
  onCheck
}) => {
  const dispatch = useDispatch();

  const onDragEnd = result => {
    const { destination, source, draggableId } = result;

    if (!destination) {
      return;
    }

    if (destination.index === source.index) {
      return;
    }

    dispatch(reorderItem(listId, draggableId, source, destination));
  };

  return (
    <DragDropContext
      onDragEnd={onDragEnd}
    >
      <Droppable droppableId={listId}>
        {provided => (
          <List {...provided.droppableProps} innerRef={provided.innerRef}>
            {items.map((item, i) =>
              <AppCSSTransition key={item.id} prefix="ListItem" timeout={200}>
                <Draggable draggableId={item.id} index={i}>
                  {provided => (
                    <ListItem
                      key={item.id}
                      item={item.text}
                      checked={item.checked}
                      onDelete={onDelete}
                      index={i}
                      itemId={item.id}
                      onCheck={onCheck}
                      draggableProps={provided.draggableProps}
                      dragHandleProps={provided.dragHandleProps}
                      innerRef={provided.innerRef}
                    />
                  )}
                </Draggable>
              </AppCSSTransition>
            )}
            {provided.placeholder}
          </List>
        )}
      </Droppable>
    </DragDropContext>
  );
};

export default DragDropList;