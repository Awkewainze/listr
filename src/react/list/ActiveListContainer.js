import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import selectActiveList from '../../redux/selectors/list/selectActiveList';
import selectActiveListId from '../../redux/selectors/list/selectActiveListId';
import ContentWrap from '../common/ContentWrap';
import NoListFallback from './NoListFallback';
import { toggleItem, deleteItem } from '../../redux/actions/list';
import NewListItemPromptContainer from './create/NewListItemPromptContainer';
import ListItem from './item/ListItem';

const ActiveListContainer = () => {
  const list = useSelector(selectActiveList);
  const activeList = useSelector(selectActiveListId);
  const dispatch = useDispatch();

  if (list == null) {
    return (
      <ContentWrap>
        <NoListFallback />
      </ContentWrap>
    );
  }

  const onToggle = itemId => {
    dispatch(toggleItem(activeList, itemId));
  };

  const onDelete = itemId => {
    dispatch(deleteItem(activeList, itemId));
  };

  return (
    <ContentWrap>
      <div className="ActiveListContainer">
        <div className="ActiveListContainer__list">
          { list.map((item, i) => <ListItem key={i} item={item.text} checked={item.checked} onDelete={onDelete} index={i} itemId={item.id} onCheck={onToggle} />) }
        </div>
        <NewListItemPromptContainer/>
      </div>
    </ContentWrap>
  );
};

export default ActiveListContainer;