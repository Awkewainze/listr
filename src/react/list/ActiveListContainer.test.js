import React from 'react';
import { shallow } from 'enzyme';
import DragDropList from './DragDropList';
import ActiveListContainer from './ActiveListContainer';
import { useSelector, useDispatch } from 'react-redux';
import NoListFallback from './NoListFallback';
import { toggleItem, deleteItem } from '../../redux/actions/list';

jest.mock('react-redux', () => {
  return {
    __esModule: true,
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
    connect: jest.fn(() => component => component)
  };
});
jest.mock('../../redux/actions/list');

const activeList = 'list12345';
const listItems = [
  {
    id: '12345',
    text: 'Test Item 12345',
    checked: false
  },
  {
    id: '54321',
    text: 'Test Item 54321',
    checked: false
  }
];

const SELECTOR_PROPS = {
  items: listItems,
  listId: activeList
};

const mockDispatch = jest.fn();

beforeEach(() => {
  useSelector.mockReset();
  useDispatch.mockReset();
  toggleItem.mockReset();
  deleteItem.mockReset();
  mockDispatch.mockReset();

  useSelector.mockImplementation(selector => SELECTOR_PROPS);
  useDispatch.mockReturnValue(mockDispatch);
});

it('matches snapshot', () => {
  expect(shallow(<ActiveListContainer />)).toMatchSnapshot();
});

it('renders fallback if no data available', () => {
  useSelector.mockImplementation(() => ({
    items: null,
    listId: null
  }));
  const wrapper = shallow(<ActiveListContainer />);
  expect(wrapper.exists(NoListFallback)).toBeTruthy();
  expect(wrapper.exists(DragDropList)).toBeFalsy();
});

it('passes correct props to action dispatchers', () => {
  const wrapper = shallow(<ActiveListContainer />);

  const firstListItem = wrapper.find(DragDropList).first();
  const onDelete = firstListItem.prop('onDelete');
  const onCheck = firstListItem.prop('onCheck');

  const testItemId = 'testItem12345';
  onDelete(testItemId);
  onCheck(testItemId);
  expect(toggleItem).toHaveBeenCalledTimes(1);
  expect(deleteItem).toHaveBeenCalledTimes(1);
  expect(mockDispatch).toHaveBeenCalledTimes(2);
  expect(toggleItem).toHaveBeenCalledWith(activeList, testItemId);
  expect(deleteItem).toHaveBeenCalledWith(activeList, testItemId);
});