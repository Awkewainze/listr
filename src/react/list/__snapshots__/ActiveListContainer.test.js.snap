// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DragDropList
  items={
    Array [
      Object {
        "checked": false,
        "id": "12345",
        "text": "Test Item 12345",
      },
      Object {
        "checked": false,
        "id": "54321",
        "text": "Test Item 54321",
      },
    ]
  }
  listId="list12345"
  onCheck={[Function]}
  onDelete={[Function]}
/>
`;
